qiskit/algorithms/minimum_eigen_solvers/adapt_qaoa.py

```diff
@@ -0,0 +1,539 @@
+        expectation = ExpectationFactory.build(
+            operator=energy_grad_op,
+            backend=self.quantum_instance,
+            include_custom=self._include_custom,
+        )
+        observable_meas = expectation.convert(StateFn(energy_grad_op, is_measurement=True))
+        ansatz_circuit_op = CircuitStateFn(wave_function)
+        expect_op = observable_meas.compose(ansatz_circuit_op).reduce()
+        return -1j * expect_op
+
+    def _test_mixer_pool(self, operator: OperatorBase):
+        self._check_problem_configuration(operator)
+        energy_gradients = []
+        for mixer in self.mixer_pool:
+            expect_op = self.compute_energy_gradient(mixer, operator, ansatz=self.ansatz)
+            # run expectation circuit
+            sampled_expect_op = self._circuit_sampler.convert(
+                expect_op, params=self.hyperparameter_dict
+            )
+            meas = sampled_expect_op.eval()
+            energy_gradients.append(meas)
+        max_energy_idx = np.argmax(np.real(energy_gradients))
+        return self.mixer_pool[max_energy_idx], np.abs(energy_gradients[max_energy_idx])
+
+    def compute_minimum_eigenvalue(
+        self,
+        operator: OperatorBase,
+        aux_operators: Optional[List[Optional[OperatorBase]]] = None,
+        iter_results=False,
+        disp=False,
+    ):
+        self.optimal_mixer_list = None
+        """Runs ADAPT-QAOA for each iteration"""
+        self._reps, self.ansatz = 1, self.initial_state  # initialise layer loop counter and ansatz
+        result_p, result = [], None
+        while self._reps < self.max_reps + 1:  # loop over number of maximum reps
+            best_mixer, energy_norm = self._test_mixer_pool(operator=operator)
+            if disp:
+                print(f"Circuit depth: {self._reps}")
+                print(
+                    f"Current energy norm | Threshold  =====> | {energy_norm} | {self.threshold} |"
+                )
+                print(f"Best mixer: {best_mixer}")
+                print()
+            if energy_norm < self.threshold:  # Threshold stoppage condition
+                if result is None:
+                    result = super().compute_minimum_eigenvalue(
+                        operator=operator, aux_operators=aux_operators
+                    )
+                break
+            self.optimal_mixer_list.append(
+                best_mixer
+            )  # Append mixer associated with largest energy gradient to list
+            self._update_ansatz(operator=operator)
+            result = super().compute_minimum_eigenvalue(
+                operator=operator, aux_operators=aux_operators
+            )
+            opt_params = result.optimal_point
+            result_p.append(result)
+            self.best_beta = list(np.split(opt_params, 2)[0])
+            self.best_gamma = list(np.split(opt_params, 2)[1])
+            if np.abs(result.optimal_value - self.ground_state_energy) < 1e-16:
+                break
```

Comment:
I believe this is not a feasible termination criterion as it assumes you know the solution (or solve the exponentially large eigenvalue problem classically) and should be removed